File: Week 2- Algorthmic warmup/6. Last digit of the sum of Fibonacci numbers/fibonacci_sum_last_digit.cpp
#include <iostream>

// int fibonacci_sum_naive(long long n) {
//     if (n <= 1)
//         return n;
//
//     long long previous = 0;
//     long long current  = 1;
//     long long sum      = 1;
//
//     for (long long i = 0; i < n - 1; ++i) {
//         long long tmp_previous = previous;
//         previous = current;
//         current = tmp_previous + current;
//         sum += current;
//     }
//
//     return sum % 10;
// }

int fibonacci_sum_fast(long long n){

/// This Function calculates modulas m of Fibonacci number n ///

// Initializations:
  int current=0;
  int previous=0;
  int tmp=0;
  int normalize=n%60; // Normalizing the fibonacci with respect to the Pisano period length
  int sum=1;
/* loop that calculates the Pisano sequence by summing the previous 2
  numbers with the breaking conditions which is another occurance of a zero
  followed by a one    */

  if (normalize==0) return 0;
  else if (normalize==1) return 1;

  else {
    current=1;
    previous=0;
    for (int i=2; i <= (normalize) ; ++i){
      tmp=current+previous;
      previous=current;
      current=tmp%10;
      sum+=current;
    }
    return sum%10;
  }

}

int main() {
    long long n = 0;
    std::cin >> n;
    std::cout << fibonacci_sum_fast(n);
}

File: Week 2- Algorthmic warmup/6. Last digit of the sum of Fibonacci numbers/fibonacci_sum_last_digit.py
# Uses python3
import sys

# def fibonacci_sum_naive(n):
#     if n <= 1:
#         return n
#
#     previous = 0
#     current  = 1
#     sum      = 1
#
#     for _ in range(n - 1):
#         previous, current = current, previous + current
#         sum += current
#
#     return sum % 10

def get_pisano_sequence(n):
    '''This Function calculates the Pisano Period'''
    arr=[0,1]
    current=1
    previous=1
    index=2
    while True:
        new=(arr[index-1]+arr[index-2])
        arr.append(new%n)
        previous=current
        current=arr[index]
        index+=1
        if (current==1 and previous==0):
            arr.pop()
            arr.pop()
            break
    return arr

def fibonacci_sum_fast(n):
    arr=get_pisano_sequence(10)
    sum=0
    for i in range(n%60+1):
        sum=sum+arr[i]
    return sum%10

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    print(fibonacci_sum_fast(n))
